src/server/components/dashboard/controllers/listsExportController.ts

```diff
@@ -0,0 +1,94 @@
+import type { NextFunction, Request, Response } from "express";
+import type { ListItem } from "server/models/types";
+import type { ListItemJsonData } from "server/models/listItem/providers/deserialisers/types";
+
+import { prisma } from "server/models/db/prisma-client";
+import { json2csv } from "json-2-csv";
+import { Readable } from "node:stream";
+import { startCase } from "lodash";
+import { format } from "date-fns";
+import { logger } from "server/services/logger";
+import { AuditEvent } from "@prisma/client";
+
+type ListWithJsonData = Array<ListItem & { jsonData: ListItemJsonData }>;
+
+export async function listsExportController(req: Request, res: Response, next: NextFunction) {
+  try {
+    const [result] = await prisma.$transaction([
+      prisma.list.findUnique({
+        where: {
+          id: Number(req.params.listId),
+        },
+        include: {
+          items: true,
+          country: true,
+        },
+      }),
+      prisma.audit.create({
+        data: {
+          type: "list",
+          auditEvent: AuditEvent.LIST_EXPORTED,
+          jsonData: {
+            listId: Number(req.params.listId),
+            userId: req.user?.id,
+            userEmail: req.user?.emailAddress,
+          },
+        },
+      }),
+    ]);
+
+    const formattedItems = formatProviderData((result?.items as ListWithJsonData) ?? []);
+
+    const output = await json2csv(formattedItems, {
+      emptyFieldValue: "none provided",
+      expandNestedObjects: false,
+      excludeKeys: [
+        "updatedAt",
+        "addressId",
+        "listId",
+        "metadata",
+        "declaration",
+        "updatedJsonData",
+        "addressCountry",
+        "type",
+        "country",
+      ],
+      preventCsvInjection: true,
+    });
+
+    const currentDate = new Date();
+    const formattedDate = format(currentDate, "yyyy-MM-dd");
+    const fileName = `${result?.country.name}-${startCase(result?.type)}-${formattedDate}.csv`;
+
+    const stream = new Readable();
+    stream.push(output);
+    stream.push(null);
+    res.set("Content-Disposition", `attachment; filename=${fileName}`);
+    res.set("Content-Type", "text/csv");
+    stream.pipe(res);
+
+    logger.info(`List ${result?.id} exported to ${fileName} by ${req.user?.emailAddress}`);
+  } catch (error) {
+    logger.error(`listsExportController error: ${error}`);
+    next("An error occurred whilst trying to export the list: ");
```

Comment:
I did but it would be good to hear your thoughts.

I had it like this previously `next(error)` but I thought the error message here would be no use to the user, so I thought I'd add something more useufl.

Reply:
That's fine. I don't think runtime errors are particularly helpful to the user but I'm referring to the colon -- It's currently "an error occurred when trying to export the list**: **" -- the : suggests there will be more text following that but I don't think there is.

Although I would probably in this case just flash an error and return them to the settings page

Reply:
Hmmm, looks like for me instead of redirecting it just downloads the error file:

```js
  } catch (error) {
    logger.error(`listsExportController error: ${error}`);
    next("There was a problem exporting the list. Please try again.");
  }
```

<img width="725" alt="Screenshot 2023-09-21 at 17 01 48" src="https://github.com/UKForeignOffice/lists/assets/1377253/6dd81dbb-04e8-4210-93b2-1a89ca248687">


I think because the anchor link has a `download` attribute.

Not sure if there's another way to do this 🤔